lib/orbit/ocs/repo/entities.ex

```diff
@@ -0,0 +1,267 @@
+defmodule Orbit.Ocs.Entities do
+  @moduledoc """
+  Responsible for processing parsed OCS messages, and writing appropriate changes
+  to the Trains and Trips in the database.
+  Also provides querying of relevant entities for the current service date.
+  """
+  require Logger
+
+  alias Orbit.Ocs.Train
+  alias Orbit.Ocs.Trip
+  alias Orbit.RailLine
+  alias Orbit.Repo
+
+  alias Orbit.Ocs.Message
+
+  alias Orbit.Ocs.Message.{
+    TschAsnMessage,
+    TschConMessage,
+    TschDelMessage,
+    TschDstMessage,
+    TschLnkMessage,
+    TschNewMessage,
+    TschOffMessage,
+    TschRldMessage,
+    TschTagMessage
+  }
+
+  import Ecto.Query
+
+  @type entities() :: [Trip.t()]
+
+  @spec query_latest(DateTime.t()) :: entities()
+  def query_latest(current_datetime \\ DateTime.utc_now()) do
+    # TODO: What filtering/sorting makes sense for this query?
+    service_date = Util.Time.service_date_for_utc_datetime(current_datetime)
+    Repo.all(from(trip in Trip, where: trip.service_date == ^service_date))
+  end
+
+  @spec to_utc(DateTime.t() | nil) :: DateTime.t() | nil
+  defp to_utc(datetime) do
+    if datetime == nil do
+      nil
+    else
+      DateTime.shift_zone!(datetime, "Etc/UTC")
+    end
+  end
+
+  @spec apply_changes(Message.t()) :: [{:ok, any} | {:error, any()}]
+  def apply_changes(%TschNewMessage{} = message) do
+    %Trip{
+      service_date: service_date(message.timestamp),
+      uid: message.trip_uid,
+      prev_uid: message.prev_trip_uid,
+      next_uid: message.next_trip_uid,
+      route: message.ocs_route_id,
+      rail_line: RailLine.from_ocs_transitline(message.transitline),
+      trip_type: message.trip_type,
+      scheduled_departure: to_utc(message.sched_dep),
+      scheduled_arrival: to_utc(message.sched_arr),
+      origin_station: message.origin_sta,
+      destination_station: message.dest_sta
+    }
+    |> Trip.changeset()
+    |> Repo.insert()
+    |> List.wrap()
+  end
+
+  def apply_changes(%TschConMessage{} = message) do
+    service_date = service_date(message.timestamp)
+    rail_line = RailLine.from_ocs_transitline(message.transitline)
+    # Apply consist to train
+    train_result =
+      %Train{
+        service_date: service_date,
+        uid: message.train_uid,
+        rail_line: rail_line,
+        cars: message.consist
+      }
+      |> Train.changeset()
+      |> Repo.insert(
+        on_conflict: {:replace, [:cars]},
+        conflict_target: [:service_date, :uid, :rail_line]
+      )
```

Comment:
question: I wonder if this should be broken out, in case we discover we want to update the consist in another part of the code? I have some recollection of us wondering if updated consists could come outside of a TSCH CON?

Reply:
So arguably the `TSCH_TAG` message also specifies a consist, which I suppose may or may not match the consist previously set by `TSCH_CON`. I'm not sure the intention/semantics of these.

At the moment I'm not really allowing the tag message to overwrite the consist. Think maybe I should change that.